loopchain/channel/channel_service.py

```diff
@@ -122,53 +120,63 @@ async def _serve():
             logging.info(f'channel_service: init complete channel: {ChannelProperty().name}, '
                          f'state({self.__state_machine.state})')
 
-        loop = MessageQueueService.loop
-        # loop.set_debug(True)
+        loop = self.__inner_service.loop
         loop.create_task(_serve())
-        loop.add_signal_handler(signal.SIGINT, self.close)
-        loop.add_signal_handler(signal.SIGTERM, self.close)
+        loop.add_signal_handler(signal.SIGINT, self.close, signal.SIGINT)
+        loop.add_signal_handler(signal.SIGTERM, self.close, signal.SIGTERM)
 
         try:
             loop.run_forever()
         except Exception as e:
             traceback.print_exception(type(e), e, e.__traceback__)
         finally:
             loop.run_until_complete(loop.shutdown_asyncgens())
+            self._cancel_tasks(loop)
+            self._cleanup()
             loop.close()
 
-            self.cleanup()
-
-    def close(self):
+    def close(self, signum=None):
+        logging.info(f"close() signum = {repr(signum)}")
         if self.__inner_service:
             self.__inner_service.cleanup()
-            logging.info("Cleanup ChannelInnerService.")
 
-        MessageQueueService.loop.stop()
+        self.__inner_service.loop.stop()
 
-    def cleanup(self):
-        logging.info("Cleanup Channel Resources.")
+    @staticmethod
+    def _cancel_tasks(loop):
+        for task in asyncio.Task.all_tasks(loop):
```

Comment:
Looks deprecated: `asyncio.Task.all_tasks`.

https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.all_tasks
> Deprecated since version 3.7, will be removed in version 3.9: Do not call this as a task method. Use the asyncio.all_tasks() function instead.

But I think that you've already knew this.
Is there any reason to use this method?

Reply:
Yes, it deprecated but work in 3.7 and loopchain will be test in `pypy3`.
So, I want to delay to use latest function.